src/main/java/de/tum/in/www1/artemis/service/RepositoryAccessService.java

```diff
@@ -59,159 +41,89 @@ public RepositoryAccessService(ParticipationAuthorizationCheckService participat
+     * for the given programming participation.
+     *
+     * @param studentParticipation The student participation.
+     * @param isLocked             True if the participation is locked, false otherwise.
+     * @return True if the user has write or reset permissions, false otherwise.
      */
-    private void checkAccessRepositoryForReset(ProgrammingExerciseParticipation programmingExerciseParticipation, boolean isStudent, ProgrammingExercise programmingExercise) {
-        boolean isOwner = true; // true for Solution- and TemplateProgrammingExerciseParticipation
-        if (programmingExerciseParticipation instanceof StudentParticipation studentParticipation) {
-            isOwner = authorizationCheckService.isOwnerOfParticipation(studentParticipation);
-        }
-        if (isStudent && programmingExerciseParticipation.isLocked()) {
-            throw new AccessForbiddenException();
-        }
-        // A tutor/instructor who is owner of the exercise should always be able to reset the repository
-        else if (!isStudent && !isOwner) {
-            // The user trying to reset the repository is at least a tutor in the course and the repository does not belong to them.
-            // This might be a tutor that resets the repository during assessment.
-            // If the student is still able to submit, don't allow for the tutor to reset the repository.
-
-            // For a regular course exercise, the participation must be locked.
-            if (programmingExercise.isCourseExercise() && !programmingExerciseParticipation.isLocked()) {
-                throw new AccessForbiddenException();
-            }
+    private boolean hasWriteOrResetPermissions(StudentParticipation studentParticipation, boolean isLocked) {
+        boolean beforeDueDate = exerciseDateService.isBeforeDueDate(studentParticipation);
+        boolean isPracticeMode = studentParticipation.isPracticeMode();
 
-            // For an exam exercise, the student must not be allowed to submit anymore.
-            // Retrieving the student via this type cast is safe because isOwner is false here which means that "programmingExerciseParticipation instanceof StudentParticipation"
-            // must have been true in the check above.
-            var optStudent = ((StudentParticipation) programmingExerciseParticipation).getStudent();
-            if (optStudent.isPresent() && programmingExercise.isExamExercise() && examSubmissionService.isAllowedToSubmitDuringExam(programmingExercise, optStudent.get(), false)) {
-                throw new AccessForbiddenException();
-            }
-        }
+        // The user has write or reset permissions if the participation is not locked and the due date has not passed yet.
+        if (beforeDueDate && !isLocked)
+            return true;
+
+        // The user has write or reset permissions if due date has passed, but the participation is in practice mode.
+        return !beforeDueDate && isPracticeMode;
```

Comment:
Small technicality here:
Even if it's practice mode, you shouldn't have access to a locked repo. I assume practice mode and locked state won't ever occur at the same time, but it should still be correctly implemented.